runtime/utils/common/src/aws/smithy/kotlin/runtime/util/net/Text.kt

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.util.net
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * Parses a `host[:port]` pair. IPv6 hostnames MUST be enclosed in square brackets (`[]`).
+ */
+@InternalApi
+public fun String.splitHostPort(): Pair<String, Int?> {
+    val lBracketIndex = indexOf('[')
+    val rBracketIndex = indexOf(']')
+    val lastColonIndex = lastIndexOf(":")
+    val hostEndIndex = when {
+        rBracketIndex != -1 -> rBracketIndex + 1
+        lastColonIndex != -1 -> lastColonIndex
+        else -> length
+    }
+
+    if (lBracketIndex == -1 && rBracketIndex != -1 ||
+        lBracketIndex != -1 && rBracketIndex == -1 ||
+        lBracketIndex > rBracketIndex
+    ) {
+        throw IllegalArgumentException("unmatched [ or ]")
+    }
+    if (lBracketIndex > 0) {
+        throw IllegalArgumentException("unexpected characters before [")
+    }
+    if (rBracketIndex != -1 && rBracketIndex != hostEndIndex - 1) {
+        throw IllegalArgumentException("unexpected characters after ]")
+    }
+
+    val host = if (lBracketIndex != -1) {
+        substring(lBracketIndex + 1 until rBracketIndex)
+    } else {
+        substring(0 until hostEndIndex)
+    }
+    if (lBracketIndex != -1 && rBracketIndex != -1 && !host.isIpv6()) {
+        throw IllegalArgumentException("non-ipv6 host was enclosed in []-brackets")
+    }
+
+    return Pair(
+        host,
+        if (hostEndIndex != -1 && hostEndIndex != length) substring(hostEndIndex + 1).toInt() else null
+    )
+}
+
+/**
+ * Validates a hostname per [RFC 1123](https://www.ietf.org/rfc/rfc1123.txt).
+ */
+@InternalApi
+public fun String.isValidHostname(): Boolean {
+    if (length !in 1..63 || !this[0].isLetterOrDigit()) {
```

Comment:
**Question**: [RFC 1123 says that](https://www.ietf.org/rfc/rfc1123.txt#:~:text=Host%20software%20MUST%20handle%20host%20names%20of%20up%20to%2063%20characters%20and%0A%20%20%20%20%20%20SHOULD%20handle%20host%20names%20of%20up%20to%20255%20characters.):

> Host software MUST handle host names of up to 63 characters and SHOULD handle host names of up to 255 characters.

Why limit to 63 in this case?

Reply:
Yeah I noticed this. It's explicitly called out to limit to the base 63 in the endpoints2 SEP, but it's worth asking those folks why they don't want to handle the 255, so I'll do that.

Reply:
It's enforced at 63 to be consistent with how Amazon (primarily S3) handles hostnames today.